netgrasp/utils/mac.py

```diff
@@ -0,0 +1,255 @@
+from __future__ import print_function
+
+import os
+import re
+import datetime
+
+TEST_WSDATA = False
+TEST_WSDATA_FILENAME = '~/.netgrasp.mac.wsdata'
+
+DB_LOOKUP_TABLE_NAME = 'macvendor'
+DB_LOOKUP_CONFIG_SOURCE_KEY = 'maclookup_source'
+DB_LOOKUP_CONFIG_UPDATED_KEY = 'maclookup_updated'
+
+LOOKUP_SOURCES = ['local', 'remote']
+
+DEFAULT_LOOKUP_SOURCE = 'remote'
+DEFAULT_LOOKUP_UPDATED = datetime.datetime(1970, 1, 1)
+DEFAULT_LOOKUP_CONFIG = dict(source=DEFAULT_LOOKUP_SOURCE, updated=DEFAULT_LOOKUP_UPDATED)
+
+WIRESHARK_ATTRIBUTION = """\
+data provided by: Wireshark - Network traffic analyzer
+                  Copyright 1998 Gerald Combs
+                  Wireshark is free software; licensed under the terms of
+                  the GNU General Public License.\
+"""
+
+
+def normalize_mac(mac):
+    """Return upper-case mac with leading 0s and ':' delimiter."""
+    mac = mac.upper()
+    pieces = re.split('[-:.]', mac)
+    mac = ':'.join(pieces)
+    if not re.match('[0-9A-F]{2}([:])[0-9A-F]{2}(\\1[0-9A-F]{2}){4}$', mac):
+        fixed = []
+        pieces = mac.split(':')
+        for piece in pieces:
+            if len(piece) == 1:
+                piece = '0' + piece
+            fixed.append(piece)
+        mac = ':'.join(fixed)
+    return mac
+
+
+def lookup_local(mac, dbfile, lockfile, debugger):
+    """Return vendor or None if not found. Raise if error."""
+    from netgrasp.database import database
+
+    debugger.debugger.debug('entering mac.lookup_local(%s)' % mac)
+
+    mac = normalize_mac(mac)
+    debugger.debug('looking up vendor for %s' % mac)
+    # Lookup vendor in db. Match mac using netmask if > 0.
+    # @TODO matching needs to be implemented; this is "placeholder" code
+    ##db = database.Database(dbfile, lockfile, debugger)
+    ##db.cursor.execute("SELECT * FROM %s WHERE ... ?" % DB_LOOKUP_TABLE_NAME, (mac,))
+    ##result = db.cursor.fetchall()
+    match = False
+    if match:
+        vendor = 'vendor'
+        debugger.debugger.debug('identified %s as %s' % (mac, vendor))
+    else:
+        vendor = None
+        debugger.debugger.info('failed to identify %s' % mac)
```

Comment:
if we're in "auto" mode (per comments above) then we likely should fallback to querying the remote API at this point; our local data may just be out-of-date.